### PR TITLE
PyTorchClassifier: use a new optimizer for the cloned classifier

### LINKED ISSUE #1579: art.estimators.classification.PyTorchClassifier: cloned classifier from `clone_for_refitting` still uses the optimizer which binds to original model's parameters
**Describe the bug**

I was using [ShadowModels](https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/art/attacks/inference/membership_inference/shadow_models.py#L58) and found that the cloned model(s) was not getting trained while strangely the original model's parameters were updated, turns out that the optimizer still binds to the original model's parameters.

Below is how an optimizer is usually created:
```
optimizer = torch.optim.SGD(model.parameters(), lr=0.01, momentum=0.9)
```
And then the same optimizer was [assigned](https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/a0c93362166e0e598545ea693d093e24b0c03655/art/estimators/classification/pytorch.py#L493) to the cloned classifier, which causes the issue.

**Suggested fix**
Currently I am using below workaround, which creates a new instance of the same type of optimizer:
```
shadow_models = ShadowModels(art_classifier, num_shadow_models=1)
for sm in shadow_models.get_shadow_models():
    new_optimizer = type(sm._optimizer)(sm.model.parameters())
    new_optimizer.load_state_dict(sm._optimizer.state_dict())
    sm._optimizer = new_optimizer
```

**System information (please complete the following information):**
 - OS: macOS
 - Python version: 3.8.10
 - ART version or commit number: 1.9.1
 - PyTorch version: 1.10.2

### PR DESCRIPTION
**Pull request by chao1995**
# Description

Fixes #1579

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [x] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Test Configuration**:
 - OS: macOS
 - Python version: 3.8.10
 - ART version or commit number: 1.9.1
 - PyTorch version: 1.10.2

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by abigailgold — CHANGES_REQUESTED** [review]
Please see my comment on not loading the optimizer's state and changing the reset method to also create a new optimizer.

File: art/estimators/classification/pytorch.py [hunk 1]
```diff
@@ -484,13 +484,18 @@ def fit_generator(self, generator: "DataGenerator", nb_epochs: int = 20, **kwarg
...
         initialization as cloned model, but without weights.
 
         :return: new estimator
         """
         model = copy.deepcopy(self.model)
-        clone = type(self)(model, self._loss, self.input_shape, self.nb_classes, optimizer=self._optimizer)
+
+        # create a new optimizer that binds to cloned model's parameters
+        new_optimizer = type(self._optimizer)(model.parameters())
+        new_optimizer.load_state_dict(self._optimizer.state_dict())
```
[the remark this answers is not in the document]
Comment (chao1995) [reply art/estimators/classification/pytorch.py:499]: Hi, thanks for the review. This line is intended to inherit any user-provided parameters like learning rate, but I'll check if there is any unintended side effects and whether there is a more explicit way to inherit the params (params_group?) without loading the whole `state_dict`.
For the resetting, wondering whether `zero_grad()` used in the training already serves the purpose, will check it out too. It may take me some time to get back on this though, apologies in advance.

File: art/estimators/classification/pytorch.py [hunk 2]
```diff
@@ -484,13 +484,22 @@ def fit_generator(self, generator: "DataGenerator", nb_epochs: int = 20, **kwarg
...
         :return: new estimator
         """
         model = copy.deepcopy(self.model)
-        clone = type(self)(model, self._loss, self.input_shape, self.nb_classes, optimizer=self._optimizer)
+
+        if self._optimizer is None:  # pragma: no cover
+            raise ValueError("An optimizer is needed to train the model, but none for provided.")
+
+        # create a new optimizer that binds to the cloned model's parameters
+        new_optimizer = type(self._optimizer)(model.parameters(), lr=self._optimizer.defaults["lr"])  # type: ignore
```
Comment (chao1995) [inline art/estimators/classification/pytorch.py:498]: Check what lr is from `self._optimizer.defaults["lr"]`. We are expecting the actual `lr` used here.

File: art/estimators/classification/pytorch.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (abigailgold) [reply art/estimators/classification/pytorch.py:499]: I think param_groups is probably the right way to go here. There may be other elements in the state besides the gradients (e.g., adaptive learning rate, weight decay, etc.) which would not be reset by zero_grad().

File: art/estimators/classification/pytorch.py (hunk 2 above)
Comment (chao1995) [reply art/estimators/classification/pytorch.py:511]: Yes, defaults contained the parameter passed in.
```
>>> import torch
>>> from torch import nn
>>> model = nn.Linear(100, 10)
>>> opt = torch.optim.Adam(model.parameters(), lr=0.1)
>>> opt.defaults
{'lr': 0.1, 'betas': (0.9, 0.999), 'eps': 1e-08, 'weight_decay': 0, 'amsgrad': False}
>>>
>>> opt2 = torch.optim.Adam(model.parameters())
>>> opt2.defaults
{'lr': 0.001, 'betas': (0.9, 0.999), 'eps': 1e-08, 'weight_decay': 0, 'amsgrad': False}
```

File: art/estimators/classification/pytorch.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (chao1995) [reply art/estimators/classification/pytorch.py]: Hi @abigailgold, thanks, removed `load_state_dict` and instead create the new optimizer with `optimizer.defaults`, which includes all the parameters provided by the user, including learning rate and weight decay etc. I suspect the defaults stay unchanged during training, but the parameters in param_groups may change (e.g. for adaptive learning rate).

File: art/estimators/classification/pytorch.py [hunk 3]
```diff
@@ -497,13 +497,21 @@ def fit_generator(self, generator: "DataGenerator", nb_epochs: int = 20, **kwarg
...
+        Create a copy of the classifier that can be refit from scratch. Will inherit same architecture, same type of
+        optimizer and initialization as the original classifier, but without weights.
 
         :return: new estimator
         """
         model = copy.deepcopy(self.model)
-        clone = type(self)(model, self._loss, self.input_shape, self.nb_classes, optimizer=self._optimizer)
+
+        if self._optimizer is None:  # pragma: no cover
+            raise ValueError("An optimizer is needed to train the model, but none for provided.")
```
Comment (beat-buesser) [inline art/estimators/classification/pytorch.py:508]:
```suggestion
            raise ValueError("An optimizer is needed to train the model, but none is provided.")
```

**Review by beat-buesser — APPROVED** [review]
Hi @chao1995 Thank you very much for you pull request fixing the cloning of PyTorch classifiers and congratulations to your first contribution to ART!

**Comment by chao1995** [comment]
Thank you @beat-buesser!
